File: 11-1.py
#!/usr/bin/env python

import re


class Monkey:
    def __init__(self, number):
        self.business = 0
        self.dest_false = None
        self.dest_true = None
        self.divisor = None
        self.items = None
        self.number = number
        self.operation = None

    def __str__(self):
        return f"<{self.__class__.__name__}  {self.number}: business: {self.business}, items: {self.items}, op: {self.operation}, div: {self.divisor}, true: {self.dest_true}, false: {self.dest_false}>"


lines = [line.strip() for line in open('11.input').readlines()]

monkeys = []

current = None

for line in lines:
    if (m := re.match("^Monkey (?P<num>[0-9]+):$", line)):
        current = Monkey(int(m['num']))
        monkeys.append(current)
    elif (m := re.match("^Starting items: (?P<items>[0-9, ]+)$", line)):
        current.items = [int(i) for i in m['items'].split(", ")]
    elif (m := re.match("^Operation: new = (?P<oper>.+)$", line)):
        current.operation = m['oper']
    elif (m := re.match("^Test: divisible by (?P<div>[0-9]+)$", line)):
        current.divisor = int(m['div'])
    elif (m := re.match("If true: throw to monkey (?P<target>[0-9]+)$", line)):
        current.dest_true = int(m['target'])
    elif (m := re.match("If false: throw to monkey (?P<target>[0-9]+)$", line)):
        current.dest_false = int(m['target'])

for round in range(20):
    for m in monkeys:
        for item in m.items:
            m.business += 1
            item = eval(m.operation, {}, {'old': item})
            item //= 3
            dest = -1
            if item % m.divisor == 0:
                dest = m.dest_true
            else:
                dest = m.dest_false
            monkeys[dest].items.append(item)
        m.items = []

s = sorted(monkeys, key=lambda m: m.business)
print(s[-1].business * s[-2].business)

File: 11-1.rb
#!/usr/bin/env ruby

class Monkey
  attr_accessor :business, :dest_false, :dest_true, :divisor, :items, :number, :operation

  def initialize(number)
    @business = 0
    @dest_false = nil
    @dest_true = nil
    @divisor = nil
    @items = nil
    @number = number
    @operation = nil
  end

  def to_s
    "<#{self.class} #{@number}: business: #{@business}, items: #{@items}, op: #{@operation}, div: #{@divisor}, true: #{@dest_true}, false: #{@dest_false}>"
  end

  alias inspect to_s
end

lines = File.read('11.input').lines.map(&:strip)

monkeys = []

current = nil

lines.each do |line|
  if (m = /^Monkey (?<num>[[:digit:]]+):$/.match(line))
    current = Monkey.new(m['num'].to_i)
    monkeys[m['num'].to_i] = current
  elsif (m = /^Starting items: (?<items>[[:digit:], ]+)$/.match(line))
    current.items = m['items'].split(', ').map(&:to_i)
  elsif (m = /^Operation: (?<oper>.+)$/.match(line))
    current.operation = m['oper']
  elsif (m = /^Test: divisible by (?<div>[[:digit:]]+)$/.match(line))
    current.divisor = m['div'].to_i
  elsif (m = /^If true: throw to monkey (?<target>[[:digit:]]+)$/.match(line))
    current.dest_true = m['target'].to_i
  elsif (m = /^If false: throw to monkey (?<target>[[:digit:]]+)$/.match(line))
    current.dest_false = m['target'].to_i
  end
end

20.times do |round|
  monkeys.each do |m|
    m.items.each do |item|
      m.business += 1
      bind = binding
      bind.local_variable_set(:old, item)
      item = eval(m.operation, bind)
      item /= 3
      dest = (item % m.divisor).zero? ? m.dest_true : m.dest_false
      monkeys[dest].items.append item
    end
    m.items = []
  end
end

monkeys.sort_by!(&:business)

print monkeys[-1].business * monkeys[-2].business, "\n"
